lib/loginator/transaction.rb

```diff
@@ -1,21 +1,125 @@
 require 'securerandom'
 
 module Loginator
-  # Methods for generating transactional metadata.
-  module Transaction
-    def self.from_json(json_str)
-      json = MultiJson.load(json_str)
-      Loginator.const_get(json['type'].capitalize).from_hash(json)
+  # A Loginator::Transaction is meant to encapsulate a single
+  # request/response cycle.
+  class Transaction
+    class << self
+      # @param [String] json JSON body of the given transaction
+      # @return [Transaction]
+      def from_json(json)
+        obj = MultiJson.load(json, symbolize_keys: true)
+        Transaction.new(filter_hash(obj))
+      end
+
+      protected
+
+      # This is the inverse functions for the instance-level
+      # filter_hash.
+      def filter_hash(hsh)
+        [[:timestamp, ->(t) { Time.at(t) }]].map do |k, f|
+          hsh[k] = f.call(hsh[k])
+        end
+        hsh
+      end
+    end
+
+    attr_accessor :path, :request, :params, :response, :status
+
+    attr_reader :uuid, :timestamp, :duration
+
+    protected
+
+    attr_writer :uuid, :timestamp, :duration
+
+    public
+
+    # Create a new Loginator::Transaction
+    # @param [Hash] opts The options to create the Loginator::Transaction
+    # @option opts [String]  :uuid UUID (SecureRandom.uuid)
+    # @option opts [Time]    :timestamp (Time.now) Beginning timestamp of transaction
+    # @option opts [Integer] :status Status returned to the requester
+    # @option opts [String]  :path  Path associated with the request
+    # @option opts [String]  :request Body of the request
+    # @option opts [Hash]    :params Parameters of the request
+    # @option opts [String]  :response Body of the response
+    # @option opts [Float]   :duration Duration of the request
+    def initialize(opts = {})
+      @uuid = opts.delete(:uuid) || SecureRandom.uuid
+      @timestamp = opts.delete(:timestamp) || Time.now
+      @duration = opts.delete(:duration)
```

Comment:
I think this is leftover from when I was doing duration differently. This can be moved back into the loop below.